### DISCUSSION TITLE
Next 13.3.0 DELETE HTTP method issue ( Edit: issue still persists in 13.4 ( April 6, 2023 - May 6, 2023 ))

### DISCUSSION THREAD

**Discussion by OatsProgramming** [vercel/next.js#48072] [Help] [answered]
### Summary

I'm creating a CRUD application to play around with the new app directory and I've encountered a problem when upgrading to Next JS 13.3.0: 

For some reason, "DELETE" does not work in the api endpoints. Prior to 13.3, I was working with 13.2.4 and I've never encountered that issue. 

I do apologize in advance if my explanation of the issue is a bit hard to understand ( I am a bit of a novice fullstack dev ), but I will try my best.

Here's my code: 

DELETE fn
```typescript
export async function DELETE(request: Request) {
    const post = await ValidateRequest(request, 'DELETE')
    if (post instanceof Error) return failedResponse(post, badRequest)
    try {
        // Identify the specific post
        const postDoc = doc(db, 'posts', post.id!)

        await deleteDoc(postDoc)
    } catch (err) {
        // On network error
        const error = err as Error
        return failedResponse(error, fetchFail)
    }
    return new Response(JSON.stringify(post), responseSuccess)
}
```

Validation method for request body
```typescript
async function ValidateRequest(request: Request, HTTP: HTTP){
    let post: PostObj;
    try {
        console.log(request)
        post = await request.json()
        console.log(post)
        switch (HTTP){
            // Check if its missing any required properties (for each method)
            .....
            case 'DELETE' : {
                if (!post.id){
                    throw new Error(`\nInvalid 'post' request body:
                        id?               ${post.id ? 'OK' : 'MISSING'}
                    `)
                }
                break;
            }
            .....
            default : {
                throw new Error('Unknown HTTP method')
            }
        }
    } catch (err) {
        // Error if missing any properties
        console.log(err)
        return err as Error
    }
    // Will return if valid 
    return post
}
```

The fetcher fn
```typescript
export async function mutatePost(method: HTTP, content: {}): Promise<void> {
    if (method === 'GET') return 
    const res = await fetch(`${url}/api/posts`, {
        method: method,
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify(content)
    })
    if (!res.ok) {
        console.log(await res.json())
    }
}
```
( Please note that the other DELETE fns are similar just different data )

With that being said, for some reason it "works" with useSWR hook. The reason why i put quotes on "works" is bc it doesn't really call the DELETE method, but it still manages to do the job. I've checked this by putting console.log() everywhere to find out what's going on, but this is for the first time i was so genuinely confused. 

That these errors ( or lack thereof if I'm using useSWR ) occured after I've upgraded from 13.2.4 to 13.3.0. 

( The error is usually just a Syntax Error for some reason: Error: Unexpected end of JSON input )

After downgrading back to 13.2.4, all went back to normal. Does anyone else have this problem?

### Additional information

```
Just in case: 


"dependencies": {
    "@next/bundle-analyzer": "^13.2.4",
    "@types/node": "18.15.0",
    "@types/react": "18.0.28",
    "@types/react-dom": "18.0.11",
    "encoding": "^0.1.13",
    "firebase": "^9.17.2",
    "framer-motion": "^10.10.0",
    "next": "^13.2.4",
    "popmotion": "^11.0.5",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "swr": "^2.1.1",
    "typescript": "4.9.5",
    "zustand": "^4.3.7"
  }
```
```


### Example

_No response_

**Comment by karim-attia** [comment]
Same issue. Had to change to method: 'POST', but I still delete in the API.

**Comment by phil-bl** [comment]
DELETE `req.body` seems to be always empty in 13.3.0

**Comment by adrianbarbic** [comment]
I'm experiencing the same issue and reverting to 13.2.4 resolves it, so yep something that has been recently introduced

**Comment by Andrei-Pene** [comment]
Thanks for Posting this. I'm having the same issue, and it worked as soon as I changed the method to POST. Hopefully it's on their list to fix

**Comment by jordonnichols** [comment]
I'm also having this issue in Next 13.4.1, works when I change the request method, seems to only be happening using DELETE method, the request body always just has nothing inside..

**Comment by karlhorky** [comment]
@OatsProgramming do you think this should be an issue (bug) instead of a discussion? The behavior you describe sounds like a bug and it may not get much visibility with the team as a discussion.

**Reply by OatsProgramming** [reply]
@karlhorky You're right and I did that with #49353 . Thought it was my code messing up which is why I posted it as a question at first.

**Reply by karlhorky** [reply]
Ok thanks!

Update: Seems your issue was closed soon afterwards in favor of this one:

- #48096

**Comment by ludovv** [comment]
Same, having this problem in 13.4.2, i just downgrade to 13.2.4

**Comment by hammadqurashi** [comment]
Same here...

**Comment by JunkyDeLuxe** [comment]
I am using v13.4.4, impossible to get the body from a DELETE REQUEST
With this error: "error SyntaxError: Unexpected end of JSON input"

In my service I do:

```
const response: AxiosResponse = await axios({
					method: 'DELETE',
					url: `/api/guest-cart`,
					data: { guestCart: cartFromLocalStorage, offerId: cartOffer.offerId },
				});
```

And my NextJs route:

```
export async function DELETE(req: NextRequest): Promise<Response> {
	console.log('delete guest cart offer');

	const { guestCart, offerId } = await req.json();

	try {
		const apiRes: Response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/order/carts-guest/front/offer`, {
			method: 'DELETE',
			body: JSON.stringify({ guestCart, offerId }),
			headers: {
				'Content-Type': 'application/json',
			},
			cache: 'no-store',
		});

		const cart: CartDetails = await apiRes.json();

		return NextResponse.json(cart);
	} catch (err) {
		return NextResponse.json({ status: 500 });
	}
}
```

**Reply by prashantchothani** [reply]
I am getting the same issue. The body in DELETE method is unable to parse and returns with 500 Error with "Unexpected end of JSON input"

**Comment by kikky7** [comment]
I was also stuck on this when moving to new App Router, since the error is very descriptive "Unexpected end of JSON input" lol.

You need to do something like this:

Button click or whatever will send this request (you can use fetch or any other solution):
```
const id = 'dummy-id';
const response = await axios.delete("/api/post", { params: { id } });
```
In this case file would look like this and in route.ts/route.js you will export all http methods you need (get, post, delete, etc.):
`/app/api/post/route.ts`

So, in your route file you do this in DELETE method:
```
export const DELETE = async (request: Request) => {
    // You should use try catch here
    
    // Get params from request
    const { searchParams } = new URL(request.url);
    // Get param named id from request (or postId or whatever name you will give to your 
    parameter)
    const id = searchParams.get("id") || "";
    
    // Do some stuff
    
    // Return new Response
};
```

Or you can do this:
Pass request like this:
```
const id = 'dummy-id';
const response = await axios.delete(`/api/post/${id}`);

// Do something with response
```

Folder structure would look like this (id is name of parameter you will send, if it's postId, then it should be [postId], etc.):
`/app/api/post/[id]/route.js`

Then in your route file you can do this (this works for GET method also):
```
export const DELETE = async (request, { params }) => {
    // You should use try catch here
    
    // Get param named id from query params
    const { id } = params;
    // Output our ID to console
    console.log("ID", id);
    
    // Do some stuff
    
    // Return new Response
};
```

**Reply by ShahariarNiloy** [reply]
i think it's kind of workaround. but it does works fine

**Comment by ImBIOS** [comment]
Why we need using req.body in DELETE method? Isn't the best practice for DELETE method is using /{id} the parameter string?

**Reply by OatsProgramming** [reply]
Best practice is to conceal data as much as possible, as often as possible when transferring data back and forth. If the task calls for a very simple DELETE for data that holds of little importance, then that's fine. 

But there are scenarios that you shouldn't reveal that data on the client side. 

Ex: you'd like to DELETE something but need to make sure that the person that's calling the method is authorized to do so. You'd need to supply credentials to do something like that; a simple /{id} won't cut it. Imagine the security risks that it poses.

**Reply by ImBIOS** [reply]
The DELETE method in HTTP is designed to request that a resource be removed. By the nature of this operation, it typically does not require a request body.

In more technical terms, the DELETE method is defined in the HTTP/1.1 specification (RFC 7231), which does not explicitly forbid or discourage sending a message (request) body. However, it does specify that a server should ignore any request body received in a DELETE request.

There are a few reasons why this is the case:

Idempotency: DELETE is an idempotent method, which means that it should have the same result, no matter how many times it is repeated. If a DELETE request were to include a body that affected how the method is carried out, it could potentially violate this property.

Simplicity: The nature of deleting a resource is such that it generally only requires the identification of that resource (which is included in the URL), and no additional data is necessary. Including a body with a DELETE request could unnecessarily complicate this operation.

Server Support: While not explicitly forbidden by the specification, many servers do not expect a DELETE request to have a body and may not support this functionality.

Despite these reasons, there might be some specific cases where you would want to include a body with a DELETE request. For example, you might need to include additional meta-information about the deletion. In such cases, it's important to know that not all servers or clients will correctly handle a DELETE request with a body. Therefore, it's generally a good practice to design your API such that a DELETE request does not require a body.

**Reply by OatsProgramming** [reply]
Do not use AI to generate answers for you. Although it has some valid points, you've completely missed the point that I was trying to make in my argument (security) and in my question (DELETE method does not work despite it being stated in the [docs](https://nextjs.org/docs/app/api-reference/file-conventions/route) ).

**Reply by PatrickBauer** [reply]
While I agree this should be fixed and that there are certainly valid use cases for this, I'm not sure I follow your argument.

The first thing is concealing data. When you send a DELETE request with a request body, than this body is as readable as the URL is, isn't it? (Unless your talking about man-in-the-middle style attacks)

Your second argument is about security. A standard security measure in this case is a HTTP request `Authorization` header. Its not really needed to send these things via the request body at all. I've only ever seen the header version to be honest. I don't know of any public APIs that I've used in the last decade that ever asked for authorization data as part of the payload/body.

**Reply by OatsProgramming** [reply]
@PatrickBauer
> The first thing is concealing data. When you send a DELETE request with a request body, than this body is as readable as the URL is, isn't it? (Unless your talking about man-in-the-middle style attacks)

Technically, yes; technically, no. URL is more visible than the body, but they are both still susceptible to middle man attacks. That's why HTTPS/SSL are a thing. Imagine having your passcode being publicly displayed in the URL: definitely breaks your trust in the site. The body has more flexibility in terms of concealing the data as compared to the URL (e.g. having your own encrypting / decrypting standards).

> Your second argument is about security. A standard security measure in this case is a HTTP request Authorization header. Its not really needed to send these things via the request body at all. I've only ever seen the header version to be honest. I don't know of any public APIs that I've used in the last decade that ever asked for authorization data as part of the payload/body.

You've made valid point: setting the credentials in the `Authorization` header is a more standardized way to do so. However, it's just as vulnerable to attacks as the request body if not using HTTPS/SSL. In the [mdn docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#examples), the practice is to set it to `Basic` and encode it to base64. Keyword, though: "encode". It's not encrypted; it can be decoded just as easily. Then again, that's just the `Basic` auth. 

> [Basic authentication](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization#basic_authentication)
>
> For "Basic" authentication the credentials are constructed by first combining the username and the password with a colon (aladdin:opensesame), and then by encoding the resulting string in [base64](https://developer.mozilla.org/en-US/docs/Glossary/Base64) (YWxhZGRpbjpvcGVuc2VzYW1l).

> ```HTTP
> Authorization: Basic YWxhZGRpbjpvcGVuc2VzYW1l
> ```

**There is no right or wrong answer** in regards to setting the credentials in the `Authorization` header or the body: it just depends on what sort of api standard you're going to implement. 

I chose to place it in the body as it more flexible for me and it's the fact that it's not standard. If you were to be in the middle man's shoes, wouldn't you look for credentials in the `Authorization` header? In addition, you can implement your own encrypting standards for credentials rather than a `Basic` base64 encoding and decrypt it on the server side.

But, as I've stated before, **there is no right or wrong answer** as it is completely up to you what standards you want to place.

**Reply by harryfoster** [reply]
Best practice or not, it would be at least nice to be able to send a response with a 405 Not Allowed when a client submits a GET or DELETE with a request body. Removing the function completely still causes the server to error when a client calls DELETE with a body. A misbehaving API client shouldn't really have the ability to cause the server to throw an error like this. 

Edit: Looks like it's fixed in 13.4.8-canary.12, for me at least.

These [docs](https://nextjs.org/docs/app/api-reference/file-conventions/route) definitely need updating too.

**Reply by OatsProgramming** [reply]
+1 to that my man

**Comment by ImBIOS** [comment]
Authentication and authorization are generally not handled within the HTTP method itself (like GET, POST, PUT, DELETE, etc.). These aspects are usually handled through other parts of the HTTP request, most commonly the headers.

One common way to handle this is by using tokens. A token is a piece of data that the server gives to a client, which the client can then provide back to the server to prove its identity or its granted permissions.

Here's how this might work:

The client authenticates with the server using a login endpoint, sending their username and password or some other form of credentials.
The server verifies these credentials and returns a token. This token is tied to the client's account and permissions.
When the client wants to perform an action that requires authorization (like deleting a resource), it includes the token in the HTTP headers of the request. Often this is done using the Authorization header.
For example, if you are using Bearer token authentication, the client would include a header like this with their DELETE request:

Authorization: Bearer TOKEN

The server can then check this token, see what permissions it grants, and decide whether or not to allow the DELETE action.
There are many types of tokens that can be used for this purpose, such as JSON Web Tokens (JWT), OAuth tokens, and others.

It's also important to note that these requests should be made over HTTPS to prevent the interception of sensitive information, like the tokens or the user's initial login credentials.

In summary, you wouldn't need to put this type of information in the request body of a DELETE request. The method itself isn't inherently insecure, but like any method, it needs to be used in combination with proper security practices.

**Reply by OatsProgramming** [reply]
I will say this again:

Do not use AI to generate answers for you. Although it has some valid points, you've completely missed the point that I was trying to make in my argument (security) and in my question (DELETE method does not work despite it being stated in the [docs](https://nextjs.org/docs/app/api-reference/file-conventions/route) ). 

For this particular argument I will provide an example: a user deleting their account. How would you do that? All have their own methods but a simple JWT and query params isn't that safe. However, to each their own.

**Comment by thiwanka778** [comment]
Use this one. 

This works for all versions on nextjs
I am using 13.4.7

     **const urlParams = new URL(req.url);
     console.log({urlParams})
    const serviceId = urlParams.searchParams.get("serviceId");**
// we can get the serviceId you passed in your request url


make sure your request url is like this http://localhost:3000/api/service?serviceId=64a0571eb5c6d3cb44e1f536
pay attention to here 

here full  DELETE function in nextjs
following function , I have not added any middleware to handle the authentication.



export async function DELETE(req: NextRequest, res: NextResponse) {
    await connectDB();


    //http://localhost:3000/api/service?serviceId=64a0571eb5c6d3cb44e1f536
// pay attention here.
     const urlParams = new URL(req.url);
     console.log({urlParams})
    const serviceId = urlParams.searchParams.get("serviceId");
 
  
    if (!serviceId) {
      return new NextResponse(
        JSON.stringify({ message: 'Service ID is required' }),
        { status: 400, headers: { 'Content-Type': 'application/json' } }
      );
    }
  
    const service = await Service.findById(serviceId);
    if (!service) {
      return new NextResponse(
        JSON.stringify({ message: "Service not found" }),
        { status: 404, headers: { 'Content-Type': 'application/json' } }
      );
    }
  

  
    const deletedService = await service.deleteOne();
  
    if (deletedService) {
      return NextResponse.json({ message: `Deleted successfully ${deletedService._id}` });
    } else {
      return new NextResponse(
        JSON.stringify({ message: 'Unexpected error. Please try again.' }),
        { status: 500, headers: { 'Content-Type': 'application/json' } }
      );
    }
  }

**Reply by harryfoster** [reply]
Now try requesting a DELETE method from a client with a request body, you'll see it still throws an error regardless. The OP's three concerns were:

1. It's a regression without a clear deprecation warning.
2. It's not documented, and infact the documentation states you can use a request body in the DELETE method.
3. It throws and error that you cannot catch cleanly to notify clients that a request body is not permissable here. 

Thankfully, it's now fixed in 13.4.8-canary.12 so you can catch it and notify the client.

**Comment by OatsProgramming** [comment, marked as answer]
Issue resolved;
Ref: #51874

**Comment by Schlumpf9** [comment]
Is it possible that this bug exists again using a **custom server** on v13.4.13?
await handle(req, res, parsedUrl); rejects the request immediatly with a 500 error "Internal Server Error" when executing a DELETE request.

Even the handler function is not reached doing a DELETE request. The following logs are not visible:
```
export default function apiHandler(req: NextApiRequest, res: NextApiResponse,) {
   logger.info('apiHandler');
   logger.info(`${req.method} ${req.url}`);
}
```
Doing an UPDATE or GET request triggers the logging. Its independent if a request body is provided or not doing a DELETE...

Downgrading to Next version 13.4.12 works just fine... I'm using the pages/api folder with a dynamic url test/[id].tsx. Probably related to #53710

**Comment by madipta** [comment]
DELETE error on Next 13.5.4 when using dynamic API route, context.params.any = undefined

**Comment by no-used** [comment]
resolved ???

**Comment by aboveyunhai** [comment]
next.js `v.13.5.5`,   `Patch` will randomly fail on production, host by Vercel, 
`{ code: "method_not_allowed", message: "Invalid HTTP method string: PATCH" }` .
localhost works fine.
Change everything to `post`  for the time being
